reactivesocket-cpp/src/Frame.cpp

```diff
@@ -351,49 +424,55 @@ std::ostream& operator<<(std::ostream& os, const Frame_KEEPALIVE& frame) {
 /// @}
 
 /// @{
-    Payload Frame_SETUP::serializeOut() {
-      auto buf = FrameBufferAllocator::allocate(FrameHeader::kSize + 3 * sizeof(uint32_t));
-
-      folly::io::Appender app(buf.get(), /* do not grow */ 0);
-      header_.serializeInto(app);
-      app.writeBE(static_cast<uint32_t>(version_));
-      app.writeBE(static_cast<uint32_t>(keepaliveTime_));
-      app.writeBE(static_cast<uint32_t>(maxLifetime_));
-      // TODO encode mime types
-      if (data_) {
-        buf->appendChain(std::move(data_));
-      }
-      return buf;
-    }
+Payload Frame_SETUP::serializeOut() {
+  auto queue = FrameBufferAllocator::allocateQueue();
+  auto buf = FrameBufferAllocator::allocate(FrameHeader::kSize + 3 * sizeof(uint32_t));
+  queue->append(std::move(buf));
+  folly::io::QueueAppender app(queue.get(), /* do not grow */ 0);
 
-    bool Frame_SETUP::deserializeFrom(Payload in) {
-      folly::io::Cursor cur(in.get());
-      if (!header_.deserializeFrom(cur)) {
-        return false;
-      }
-      // TODO support metadata
-      assert((header_.flags_ & FrameFlags_METADATA) == 0);
-      try {
-        version_ = cur.readBE<uint32_t>();
-        keepaliveTime_ = cur.readBE<uint32_t>();
-        maxLifetime_ = cur.readBE<uint32_t>();
-      } catch (...) {
-        return false;
-      }
-      // TODO decode mime types
-      auto totalLength = cur.totalLength();
-      if (totalLength > 0) {
-        cur.clone(data_, totalLength);
-      } else {
-        data_.reset();
-      }
-      return true;
-    }
+  header_.serializeInto(app);
+  app.writeBE(static_cast<uint32_t>(version_));
+  app.writeBE(static_cast<uint32_t>(keepaliveTime_));
+  app.writeBE(static_cast<uint32_t>(maxLifetime_));
+  // TODO encode mime types
+  if (header_.flags_ & FrameFlags_METADATA) {
+    app.insert(metadata_->serialize());
+  }
+  if (data_) {
+    app.insert(std::move(data_));
+  }
+  return queue->move();
+}
 
-    std::ostream& operator<<(std::ostream& os, const Frame_SETUP& frame) {
-      return os << frame.header_ << "(<"
-             << (frame.data_ ? frame.data_->computeChainDataLength() : 0)
-             << ">)";
-    }
+bool Frame_SETUP::deserializeFrom(Payload in) {
+  folly::io::Cursor cur(in.get());
+  if (!header_.deserializeFrom(cur)) {
+    return false;
+  }
+  try {
+    version_ = cur.readBE<uint32_t>();
+    keepaliveTime_ = cur.readBE<uint32_t>();
+    maxLifetime_ = cur.readBE<uint32_t>();
+  } catch (...) {
+    return false;
+  }
+  // TODO decode mime types
```

Comment:
I can fix this issue since the case I mentioned was already badly broken.  But can you disable reading metadata for setup frames until I do.  Or fix both in this diff.  Your call.